sandbox/grist/predicate_formula.py

```diff
@@ -63,6 +64,67 @@ def parse_predicate_formula_json(formula):
   'None': None,
 }
 
+
+def process_renames(formula, collector, renamer):
+
+  """
+  Given a predicate formula, a collector and a renamer, rename all references in the formula
+  that the renamer wants to rename. This is used to automatically update references in an ACL
+  or dropdown condition formula when a column it refers to has been renamed.
+
+  The collector should be a subclass of TreeConverter that collects related NamedEntity's and
+  stores them in the field "entities". See acl._ACLEntityCollector for an example.
+
+  The renamer should be a function taking a NamedEntity as its only argument. It should return 
+  a new name for this NamedEntity when it wants to rename this entity, or None otherwise.
+  """
+  patches = []
+  # "$" can be used to refer to "rec." in Grist formulas, but it is not valid Python.
+  # We need to replace it with "rec." before parsing the formula, and restore it back after
+  # the surgery.
+  # Keep the dollar replacer object, so that later we know how to restore properly.
+  dollar_replacer = get_dollar_replacer(formula)
+  formula_nodollar = dollar_replacer.get_text()
+  try:
+    atok = asttokens.ASTTokens(formula_nodollar, tree=ast.parse(formula_nodollar, mode='eval'))
+    collector.visit(atok.tree)
+  except SyntaxError:
+    # Don't do anything to a syntactically wrong formula.
+    return formula
+  except ValueError as e:
+    if str(e).startswith("Unsupported syntax"):
+      return formula
+
+  for subject in collector.entities:
+    new_name = renamer(subject)
+    if new_name is not None:
+      patches.append(textbuilder.make_patch(
+        dollar_replacer.get_text(), subject.start_pos, subject.start_pos + len(subject.name), new_name))
+
+  new_formula = textbuilder.Replacer(dollar_replacer, patches)
+  new_formula_text = new_formula.get_text()
+
+  # Find all "rec." in the processed formula.
+  rec_occurrences = []
+  cursor = 0
+  while True:
+    next_occurrence = new_formula_text.find("rec.", cursor)
+    if next_occurrence == -1:
+      break
+    cursor = next_occurrence + 4 # "rec." has 4 characters
+    rec_occurrences.append(next_occurrence)
+
```

Comment:
This works, but it feels a little too manual and verbose. Wouldn't you rather use the `re` module?
```python
rec_occurrences = (m.start() for m in re.finditer("rec.", new_formula_text))
```
(Note the use of round brackets `()` and not square `[]` so that we can benefit from lazy evaluation in line 119.)

Reply:
Beautiful code. One small gotcha: pattern should be `r"rec\."`.